### ISSUE TITLE
Error evaluating injectedJavaScript: This is possibly due to an unsupported return type. Try adding true to the end of your injectedJavaScript string.

### ISSUE THREAD

**Issue by EgidioCaprino** [react-native-community/react-native-webview#341]
Hello,

We're trying to migrate from `react-native` `WebView`. Replacing `import { WebView } from 'react-native'` with `import { WebView } from 'react-native-webview'` gives us this error when the component is mounted:

```
Error evaluating injectedJavaScript: This is possibly due to an unsupported return type. Try adding true to the end of your injectedJavaScript string.
```

I've tried to add `; true` or `; return true` at the end of the injected script but we still get the same error.

Do you know how we can fix it or identify the underlying problem?

**Comment by RobinUS2** [comment]
What worked for us was this:
```
injectJavaScript(
                "window.callSomeFunction(); void(0);",
              )
```

**Comment by EgidioCaprino** [comment]
Thank you so much @RobinUS2 ❤️

**Comment by EgidioCaprino** [comment]
I've noticed that if I stop pause between one call to `webView.injectJavaScript` and the next one with the debugger it does not throw any error

**Comment by EgidioCaprino** [comment]
Throttling the calls to `injectJavaScript` seems to have fixed the problem.

```js
this.injectJavaScript = throttle(this.injectJavaScript.bind(this), 500);
```

**Comment by matthewfbenjamin** [comment]
@EgidioCaprino's solution worked for me. I had to stop using the `injectedJavaScript` prop and used the method instead to inject the same string.

**Comment by EgidioCaprino** [comment]
This problem still occurs for us 😢

**Comment by legion-zver** [comment]
See your JS code carefully, this error is due to the fact that WebView can not execute JS code!

**Comment by EgidioCaprino** [comment]
The same code was working with `WebView` imported from `react-native` 😕

**Comment by Titozzz** [comment]
Android or iOS ?

**Comment by meshaabi** [comment]
just upgraded from 2.10 to the current latest and started running into this. The injection seems a wee bit more unreliable in general after the upgrade.

**Comment by Titozzz** [comment]
This seems to happen when you inject code too early before webview is loaded, but I'm not sure

**Comment by sillysachin** [comment]
Facing it with some form submission . 
**Working -**        
`            const jsCodeHello = 'alert("hello")';
`
**Not Working -**       
```
const jsCode = `
     document.getElementById("name").value = "${name}"; 
     document.f1.submit(); 
`; 
```

**Comment by sillysachin** [comment]
```
        const jsCode = `
            setTimeout(() => {
                // document.getElementById("name").value = "${name}"; 
                document.f1.submit(); 
            }, 100);
        `;
```

**Comment by alexlevy0** [comment]
Same issue when i'm using template string even with a big timer :/

**Comment by Titozzz** [comment]
When I inject stuff inside the webview i'll always wrap it inside a function and instantly execute it. It allows me to avoid wrong return types:

```js
(function(lol){
	//do stuff
})()
```

**Comment by AndrewHenderson** [comment]
@legion-zver is right. Try running that same code in a Web browser console and verify that it can execute. There is likely a minor syntax error.

**Comment by fangasvsass** [comment]
i have the same issue , The same code was working with WebView imported from react-native,Who has a good solution?

**Comment by andrevvalle** [comment]
This works for me!

`this.webref.injectJavaScript(this.onLoadGetStore);`

![Screen Shot 2019-04-12 at 14 37 09](https://user-images.githubusercontent.com/7549986/56055715-97e9b500-5d30-11e9-82cf-e7fad9d889ff.png)

**Comment by brandonmoon** [comment]
This is really late, but for anyone else stuck on this issue and having problems finding a solution, for me the issue was that in the original react-native WebView I could just call `postMessage()`, but this library changed it to `window.ReactNativeWebView.postMessage()`. Since `postMessage` is now undefined, your existing code will throw an error, and the new library responds with the generic (and very unhelpful) error message: 

> ```Error evaluating injectedJavaScript: This is possibly due to an unsupported return type. Try adding true to the end of your injectedJavaScript string.``` 

You will see this message for _any_ JS error thrown by the webview. By updating your `postMessage()` calls you should be able to get things working again. Good luck!

TL;DR => Replace `postMessage()` calls with `window.ReactNativeWebView.postMessage()` to fix this error.

**Comment by Iamhere** [comment]
In my case I replaced `injectedJavaScript` prop with injectJavaScript() method and it fixed the error.

**Comment by Slapbox** [comment]
@sillysachin's solution worked for me, and even with a large injection on a very slow simulator, using `0` for the amount of time works just as well as `100`.

**Comment by davidcort** [comment]
This bug is present in iOS <= 9.2. For recent releases this bug is not present.

**Comment by KangYoosam** [comment]
using `setTimeout` solved my problem as described here.
https://github.com/react-native-community/react-native-webview/issues/341#issuecomment-466639820

**Comment by Juanjo4U** [comment]
try and catch works fine, if you are adding an event to an element that does not exist you are gonna get (Error evaluating injectedJavaScript: This is possibly due to an unsupported return type. Try adding true to the end of your injectedJavaScript string.
#)
so if u use try and catch if the element does not exist u are gonna get whatever you put in catch, but if element exist you are gonna get what u put on try, don't forget to add a true at the end of the injection.

for example I used this: 
const INJECTED_JS = `
      try{
            document.querySelector('form[data-refresh-athlete=true]').addEventListener('submit', _ => {
                window.ReactNativeWebView.postMessage(
                    JSON.stringify({
                        requireUserFecth: true,
                    }),
                )
            })
        } catch(e) {
            window.ReactNativeWebView.postMessage(
                JSON.stringify({
                    injected: false,
                }),
            )
        }
        true;
`

**Comment by shinoaki-tatsuya** [comment]
react-native-webview outputs the message "Please add true to the end of the inserted JavaScript string" regardless of the error that occurred in `injectedJavaScript`.
The latest version of react-native-webview adds a dump of the actual error that occurred at the end of the message to help identify the real cause.
https://github.com/react-native-community/react-native-webview/blob/5e8b4d5c2b08b982b1f80ecd34160597ec482a4f/ios/RNCWebView.m#L999

I think that there are many cases where `WKWebView was invalidated` has occurred.

**Comment by GreatAuk** [comment]
> This seems to happen when you inject code too early before webview is loaded, but I'm not sure

It is not because inject code too early before webview is loaded, the reason is inject code too early before `window` is loaded
this [solution](https://github.com/react-native-community/react-native-webview/issues/341#issuecomment-482661463) is worked。

**Comment by harking** [comment]
In our error, we were trying to inject during `componentWillUnmount` to try and clean up some resources the webview was using. Removing this fixed it.

**Comment by Leshgan** [comment]
using prop `injectedJavaScriptBeforeContentLoaded` instead `injectedJavaScript` helped for me

**Comment by GreatAuk** [comment]
> This seems to happen when you inject code too early before webview is loaded, but I'm not sure

issue still （some time）
this works for me
```
    `
      try {
        window.onload = function() {
          callFunc()
        }
        if(document.readyState === 'complete') {
         callFunc()
        }
      } catch (err) {
        console.error(err)
      }
      true;
      `
      // it's may be call function for twice
```

**Comment by svpease** [comment]
Had this problem, realized it was a function in my code that was looping indefinitely within the webview. So don't forget to check if that's happening.

**Comment by Nesh108** [comment]
Same problem! It flickers quite heavily on iOS too.

**Comment by fondue-tech** [comment]
Amazing still issue after 3 years. Crazy.

**Comment by ElectricCoffee** [comment]
Yeah I'm having this issue as well.

I've noticed it only happens when I try to call `window.ReactNativeWebView.postMessage`

**Comment by emrahyurttutan** [comment]
> Fixed using a timer.
> 
> ```
> const jsCode = `
>     setTimeout(() => {
>         document.getElementById("name").value = "${name}"; 
>         document.f1.submit(); 
>     }, 100);
> `;
> ```

`const jsCode = `
        document.getElementById("name").value = "${name}"; 
        document.f1.submit(); 
`;`

` <WebView
          javaScriptEnabledAndroid
          javaScriptEnabled
          injectedJavaScriptBeforeContentLoaded={jsCode}
        />`

**Comment by ElectricCoffee** [comment]
> > Fixed using a timer.
> > ```
> > const jsCode = `
> >     setTimeout(() => {
> >         document.getElementById("name").value = "${name}"; 
> >         document.f1.submit(); 
> >     }, 100);
> > `;
> > ```
> 
> this could be an alternative
> 
> ```
> <WebView 
> javaScriptEnabledAndroid
> javaScriptEnabled
> injectedJavaScriptBeforeContentLoaded={'document.getElementById("name").value = "${name}";document.f1.submit();'}
> />
> ```

Except it doesn't because you didn't interpolate the string in the injected JS. Replace ' with ` and you're golden.

**Comment by sz423242122** [comment]
> Fixed using a timer.
> 
> ```
> const jsCode = `
>     setTimeout(() => {
>         document.getElementById("name").value = "${name}"; 
>         document.f1.submit(); 
>     }, 100);
> `;
> ```

Thanks~ It fixed my problem！
